src/frontend/xgboost/xgboost_json.h

```diff
@@ -0,0 +1,376 @@
+   * \param key the JSON key for this output
+   * \param value the value to be assigned
+   * \param output reference to object to which the value should be assigned
+   */
+  template <typename ValueType>
+  bool assign_value(const std::string &key,
+                    ValueType &&value,
+                    ValueType &output);
+  template <typename ValueType>
+  bool assign_value(const std::string &key,
+                    const ValueType &value,
+                    ValueType &output);
+
+ private:
+  /* \brief the delegator which delegated parsing responsibility to this handler */
+  std::weak_ptr<Delegator> delegator;
+  /* \brief the JSON key for the object currently being parsed */
+  std::string cur_key;
+};
+
+/*! \brief JSON handler that ignores all delegated input*/
+class IgnoreHandler : public BaseHandler {
+ public:
+  using BaseHandler::BaseHandler;
+  bool Null();
+  bool Bool(bool b);
+  bool Int(int i);
+  bool Uint(unsigned u);
+  bool Int64(int64_t i);
+  bool Uint64(uint64_t u);
+  bool Double(double d);
+  bool String(const char *str, std::size_t length, bool copy);
+  bool StartObject();
+  bool Key(const char *str, std::size_t length, bool copy);
+  bool StartArray();
+};
+
+/*! \brief base handler for updating some output object*/
+template <typename OutputType> class OutputHandler : public BaseHandler {
+ public:
+  /*! 
+   * \brief construct handler to be added to given delegator's stack
+   * \param parent_delegator pointer to Delegator for this handler
+   * \param output the object to be modified during parsing
+   */
+  OutputHandler(std::weak_ptr<Delegator> parent_delegator,
+                OutputType &output)
+      : BaseHandler{parent_delegator}, m_output{output} {};
+  OutputHandler(std::weak_ptr<Delegator> parent_delegator,
+                OutputType &&output) = delete;
+
+ protected:
+  /* \brief the output value constructed or modified during parsing */
+  OutputType &m_output;
+};
+
+/*! \brief handler for array of objects of given type*/
+template <typename ElemType, typename HandlerType = BaseHandler>
+class ArrayHandler : public OutputHandler<std::vector<ElemType>> {
+ public:
+  using OutputHandler<std::vector<ElemType>>::OutputHandler;
+  bool Bool(ElemType b) {
+    this->m_output.push_back(b);
+    return true;
+  }
+  template <typename ArgType, typename IntType = ElemType>
+  typename std::enable_if<std::is_integral<IntType>::value, bool>::type
+  store_int(ArgType i) {
+    this->m_output.push_back(static_cast<ElemType>(i));
+    return true;
+  }
+
+  template <typename ArgType, typename IntType = ElemType>
+  typename std::enable_if<!std::is_integral<IntType>::value, bool>::type
+  store_int(ArgType i) {
+    return false;
+  }
+
+  bool Int(int i) { return store_int<int>(i); }
+  bool Uint(unsigned u) { return store_int<unsigned>(u); }
+  bool Int64(int64_t i) { return store_int<int64_t>(i); }
+  bool Uint64(uint64_t u) { return store_int<uint64_t>(u); }
```

Comment:
```suggestion
  bool Int(int i) override { return store_int<int>(i); }
  bool Uint(unsigned u) override { return store_int<unsigned>(u); }
  bool Int64(int64_t i) override { return store_int<int64_t>(i); }
  bool Uint64(uint64_t u) override { return store_int<uint64_t>(u); }
```

Likewise, use `override` keyword whenever a function of a derived class overrides a virtual function of the base class.